pkg/orchestrator/selection/ranking/available_capacity.go

```diff
@@ -0,0 +1,151 @@
+	// Get dynamic weights based on job requirements
+	jobResources, err := job.Task().ResourcesConfig.ToResources()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get job resources: %w", err)
+	}
+	cpuWeight, memoryWeight, diskWeight, gpuWeight := dynamicWeights(jobResources)
+
+	// Initialize variables to track maximum weighted capacities
+	var maxWeightedAvailableCapacity, maxQueueUsedCapacity float64
+	weightedAvailableCapacities := make(map[string]float64, len(nodes))
+	weightedQueueCapacities := make(map[string]float64, len(nodes))
+
+	// Calculate weighted capacities for each node and determine the maximum values
+	for _, node := range nodes {
+		weightedAvailableCapacity := weightedCapacity(node.ComputeNodeInfo.AvailableCapacity, cpuWeight, memoryWeight, diskWeight, gpuWeight)
+		weightedQueueUsedCapacity := weightedCapacity(node.ComputeNodeInfo.QueueUsedCapacity, cpuWeight, memoryWeight, diskWeight, gpuWeight)
+
+		weightedAvailableCapacities[node.NodeID] = weightedAvailableCapacity
+		weightedQueueCapacities[node.NodeID] = weightedQueueUsedCapacity
+
+		if weightedAvailableCapacity > maxWeightedAvailableCapacity {
+			maxWeightedAvailableCapacity = weightedAvailableCapacity
+		}
+		if weightedQueueUsedCapacity > maxQueueUsedCapacity {
+			maxQueueUsedCapacity = weightedQueueUsedCapacity
+		}
+	}
+
+	// Rank nodes based on normalized weighted capacities
+	ranks := make([]orchestrator.NodeRank, len(nodes))
+
+	for i, node := range nodes {
+		weightedAvailableCapacity := weightedAvailableCapacities[node.NodeID]
+		weightedQueueUsedCapacity := weightedQueueCapacities[node.NodeID]
+
+		// Calculate the ratios of available and queue capacities
+		availableRatio := 0.0
+		queueRatio := 0.0
+
+		if maxWeightedAvailableCapacity > 0 {
+			availableRatio = weightedAvailableCapacity / maxWeightedAvailableCapacity
+		}
+		if maxQueueUsedCapacity > 0 {
+			queueRatio = weightedQueueUsedCapacity / maxQueueUsedCapacity
+		}
+
+		// Normalize the ratios to the rank range
+		normalizedAvailableRank := availableRatio * float64(maxAvailableCapacityRank)
+		normalizedQueueRank := (1 - queueRatio) * float64(maxQueueCapacityRank)
+
+		// Calculate the final rank, higher available capacity and lower queue used capacity should give a higher rank
+		rank := normalizedAvailableRank + normalizedQueueRank
+
+		// Ensure the rank is within the desired range
+		rank = math.Max(rank, float64(orchestrator.RankPossible))
+		rank = math.Min(rank, maxAvailableCapacityRank+maxQueueCapacityRank)
+
+		// Assign rank and reason to the node
+		ranks[i] = orchestrator.NodeRank{
+			NodeInfo: node,
+			Rank:     int(rank),
+			Reason: fmt.Sprintf(
+				"Ranked based on available capacity %s and queue capacity %s",
+				node.ComputeNodeInfo.AvailableCapacity.String(), node.ComputeNodeInfo.QueueUsedCapacity.String()),
+			Retryable: true,
+		}
+		log.Ctx(ctx).Trace().Object("Rank", ranks[i]).Msg("Ranked node")
+	}
+
+	return ranks, nil
```

Comment:
Let's return the ranks in sorted order, will remove one step from the tests and make it easier for humans to parse log lines.

Reply:
answered in https://github.com/bacalhau-project/bacalhau/pull/4002#discussion_r1605968664